Extended selection commands

Zacrain opened the pull request:
Hey there,

I extended the selection command ('/npc select') as follows:

It is now possible to select the npc based on either:

- the id (index in the set of currently available NPCs) of the NPC
- the name of the NPC
- the UUID of the NPC

The commands are extended by corresponding literals in order to correctly distinguish between those selection methods.
That looks like this:

- Select by id:
  - `/npc select id <id>`
- Select by name: 
  - `/npc select name <name>`
- Select by UUID: 
  - `/npc select uuid <uuid>`

Note, that for names containing whitespaces (e.g. space) the name has to be enclosed in quotation marks.
Also in case of more than one NPC with that name (e.g. two NPCs named 'Foobar') an error message will be displayed, informing the user that there is more than one NPC with that particular name (and the selection thus not possible).

Examples:

- `/npc select id 3`
- `/npc select name "Foo bar"`
- `/npc select name Baz`
- `/npc select uuid f81d4fae-7dec-11d0-a765-00a0c91e6bf6`

Furthermore I added tab completion suggestions for each of those selection methods. In case of names as selection method, the names are automatically enclosed within quotation marks.

Of course I also added translation keys (but provided text just for en_us).

Don't be confused about the huge amount of differences. Something weird happened there with the remapped files being pushed. I am new to minecraft modding, this is the first time I forked another project in order to contribute.
I mainly modified 'NpcCommand.java'. Added keys in en_us.json and in permissions.toml.

I find those additional capabilities very useful, especially if you use mc-functions or other automated scripts ingame to change the behaviour of the NPCs dynamically. This is because the id (list index) itself is not sufficient to distinguish uniquely and universally between the NPCs, since the order depends on how the chunks, resp. the NPCs are loaded into the 'TATERZEN_NPCS' set. And that again heavily depends on the player position and movement.

I built and tested that with fabric. Can't say anything about forge. I also did not update the wiki (yet) but will do that as soon as I've got the time.

Best regards,

Zacrain

samolego commented:
First off, thank you for taking your time and contributing to the project! Welcome to the modding community :).
I have suggested some changes, but the[ first 2 commits](https://github.com/samolego/Taterzens/pull/93/commits/1e0cd5a8066f9875112b3175aeab1497630c1977) makes me a little bit worried. Could you please exclude it?

samolego commented:
> I find those additional capabilities very useful, especially if you use mc-functions or other automated scripts ingame to change the behaviour of the NPCs dynamically. This is because the id (list index) itself is not sufficient to distinguish uniquely and universally between the NPCs, since the order depends on how the chunks, resp. the NPCs are loaded into the 'TATERZEN_NPCS' set. And that again heavily depends on the player position and movement.

Thinking about this now, `LinkedHashSet<TaterzenNPC>` could be migrated to `LinkedHashMap<UUID, TaterzenNPC>`.

Zacrain commented:
> First off, thank you for taking your time and contributing to the project! Welcome to the modding community :). I have suggested some changes, but the[ first commit](https://github.com/samolego/Taterzens/pull/93/commits/1e0cd5a8066f9875112b3175aeab1497630c1977) makes me a little bit worried. Could you please exclude it?

Thank you! :)

Yes, I removed that strange commit. Should be fine now.

I don't see where you suggested changes. Can you point me to them?

Zacrain commented:
> I also did not update the wiki (yet) but will do that as soon as I've got the time.

I have just extended the documentation for the new selection features.

samolego commented:
I've added comments to the commits, do they not show up for you? 🤔

samolego commented:
:ohno: I wanted to "request for changes" but clicked approve instead :sweat_smile:

samolego commented on common/src/main/java/org/samo_lego/taterzens/commands/NpcCommand.java:
```diff
@@ -198,6 +215,106 @@ private static int selectTaterzenById(CommandContext<CommandSourceStack> context
+        for (TaterzenNPC npcIt : TATERZEN_NPCS) {
+            if (npcIt.getName().getString().equals(name)) {
+                taterzen = npcIt;
+                count++;
```
Then you could also use `boolean found` instead of `int count`

Zacrain replied:
Good idea for saving loop iterations! I made changes according to your suggestion. But I did not use a boolean, since using the count variable (which can become greater than 1 and, theoretically, take even negative values) makes such a boolean redundant.

Zacrain commented on common/src/main/resources/data/taterzens/lang/en_us.json:
```diff
@@ -100,8 +100,12 @@
   "taterzens.tooltip.exit": "Exit",
   "taterzens.error.invalid.url": "URL %s is malformed.",
   "taterzens.error.invalid.text": "The text nbt you entered is invalid.",
+  "taterzens.error.invalid.uuid": "The uuid you entered is invalid.",
```
Cool idea, didn't knew that. Changed it. 😄

Zacrain commented:
> ~~I've added comments to the commits, do they not show up for you? 🤔 ~~
> 
> Seems like I've messed it :P

Did not see them, did not get a notification or something like that. Had to go through the commits manually and look for open conversations.
(I used gitlab way more often than github. 😁  )

Zacrain commented:
> :ohno: I wanted to "request for changes" but clicked approve instead 😅

Fine with me. 😁 I incorporated the changes you suggested. 😉 

I also took the liberty to extend the sound capabilities of the Taterzens. It is now possible to change their sounds individually. They still can get initialized with global default values from the config file, but individual changes will take precedence. Saving and loading individual sound changes of Taterzens works fine as well.

I have also added a documentation for the new sound command, where the new feature is detailed. See [docs/getting_started/sounds.md ](https://github.com/samolego/Taterzens/pull/93/commits/72a5e0150d6e154524fa77055d01396e46f4ebfa#diff-a99a3703a5dad9cbc37a2acf285a49982b2ffb03ab4ebbf4715d759e09524f7b).

samolego commented:
GUI is created automatically for each command node with the help of recursion magic :).

samolego commented:
Do you consider adding anything else (I'd advise a new PR for more features) or is this ready for merging?

Zacrain commented:
> GUI is created automatically for each command node with the help of recursion magic :).

Uhh cool. 😮 Good to know.

> Do you consider adding anything else (I'd advise a new PR for more features) or is this ready for merging?

It is ready for merging. :) 

I still have a few things on my agenda, which I would like to add to this project, but they are unrelated and I'll do a new pull request for them.
